tenzir/threatbus
# Improve message distribution performance

###  :notebook_with_decorative_cover: Description

Improve (retro-matching) performance.

#### VAST Bridge:

- use asynchronous tasks to query VAST concurrently
- add configurable `ulimit` to control task count

#### Threat Bus:

- use [multiprocessing.JoinableQueue](https://docs.python.org/3.8/library/multiprocessing.html#multiprocessing.JoinableQueue) for message passing
- use `select` system call (possible via JoinableQueue) to speed up message distribution
- add `StoppableWorker` base-class for background threads
- control shutdown behavior of plugins

###  :memo: Checklist

- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.

### :dart: Review Instructions

*File-by-file.*

Test this as follows:

- setup a virtual environment, somewhere on your computer, via `virtualenv venv`
- activate it `source venv/bin/activate`
- run `make dev-mode`
- write a config that activates all plugins (install broker, CIF, MISP, RabbitMQ ... locally first, see READMEs)
- run Threat Bus via `venv/bin/threatbus -c config.yaml`
- Send around IoCs: create / delete in MISP, verify they make it to Zeek / VAST, browse sites that trigger these IoCs, watch the sightings being collected and sent back.
